File: baekjoon/src/binarySearch/B_22871.java
package binarySearch;

import java.util.*;
import java.io.*;

public class B_22871 {
	static int N;
	static long[] a;
	static long[] dp;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		N = Integer.parseInt(br.readLine());
		a = new long[N];
		dp = new long[N];

		StringTokenizer st = new StringTokenizer(br.readLine());
		for(int i=0;i<N;i++) 
			a[i] = Long.parseLong(st.nextToken());

		System.out.println(isCross(0));
	}

	static long isCross(int n) {
		if(n == N - 1)
			return 0;

		if(dp[n] == 0) {
			dp[n] = Long.MAX_VALUE;
			for(int i=n+1;i<N;i++) 
				dp[n] = Math.min(dp[n], Math.max((i - n) * (Math.abs(a[i] - a[n]) + 1), isCross(i)));
		}
		
		return dp[n];
	}

}

File: src/kotlin/dp/B_22871.kt
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

lateinit var arr: LongArray
lateinit var dp: LongArray
val max = Long.MAX_VALUE
var n = 0
fun main() = with(System.`in`.bufferedReader()) {
    n = readLine().toInt()
    arr = readLine().split(" ").map { it.toLong() }.toLongArray()
    dp = LongArray(n) { max }
    dp[n - 1] = 0
    print(solution(0))
}

fun solution(idx: Int): Long {
    if (dp[idx] != max) {
        return dp[idx]
    }

    for (i in idx + 1..<n) {
        val value = (i - idx) * (1 + abs(arr[idx] - arr[i]))
        dp[idx] = min(dp[idx], max(solution(i), value))
    }
    return dp[idx]
}
